runtime/lua/vim/diagnostic.lua

```diff
@@ -1179,6 +1242,24 @@ function M.config(opts, namespace)
   end
 end
 
+--- Execute a given function now if the given buffer is already loaded or once it is loaded later.
+---
+---@param bufnr integer Buffer number
+---@param fn fun()
+local function once_buf_loaded(bufnr, fn)
+  if api.nvim_buf_is_loaded(bufnr) then
+    fn()
+  else
+    api.nvim_create_autocmd('BufRead', {
+      buffer = bufnr,
+      once = true,
+      callback = function()
+        fn()
+      end,
+    })
+  end
+end
```

Comment:
This is useful in particular for `vim.diagnostic.set(vim.diagnostic.fromqflist())` use case.

When using `:make` or similar it's normal to end up with unloaded buffers associated with quickfix items if you didn't edit that buffer before running `:make`, in which case it would not create extmarks and you wouldn't be able to jump between diagnostics.

Builtin diagnostic handlers currently do nothing for unloaded buffers, meaning you won't get underlines, virtual text etc.
Ideally they should also be changed to use this function. I planned to do that in a separate follow-up PR though.